Updated ConfirmDetailsAreValidForModulusCheck.cs

sourcejon opened the pull request:
Added the ability to control the return Boolean when there are no SortCode Modulus Weight Mappings available. This will default to 'isNotFoundValid = true' so there are no breaking changes.

sourcejon committed:
Updated ConfirmDetailsAreValidForModulusCheck.cs

Added the ability to control the return Boolean when there are no SortCode Modulus Weight Mappings available. This will default to 'isNotFoundValid = true' so there are no breaking changes.

pauldambra commented:
Hey, thanks for the PR.

I'm not clear on the use-case. My understanding of the modulus checking process is that if there is no weight mapping then the result is `valid` which this library represents as true...

This change would allow you to return false which doesn't follow the algorithm.

(I'm a bit confused as to why the diff shows your file as entirely new too but that's probably neither here nor there)

sourcejon committed:
Updated ModulusChecker.cs

Allows you to optionally specify the behavior when sort code weight mappings are not located.

sourcejon commented:
Good morning and thank you for your fast response.

Sorry I am a little new with github, so that may be why the files are not showing changes correctly.

You are indeed correct about the modulus checking process.

We use this in house and we have modified it such that it gives you the option when checking a bank account to specify a not found sort code as a fail. This is often useful in the real world as humans are useless at inputting data.

This code change is non breaking, so everything would default to the current behavior you described this just gives people the flexibility to fail a check if mappings are not found.

Cheers,
Jon

pauldambra commented:
Don't worry about being new to Github we can sort out correctness if we come to merge it :)

I'm a bit allergic to boolean parameters they tend to mean there's really a missing method or concept...

Since I first wrote this library the ModulusChecking process has changed to include a check against the extended sort code directory as the first step to check if it's a valid sort code.
 
I'm a bit worried about using a Boolean to mean two different things... If you get false is it false or false, you know?

I've thought for a while that we could do with an `explain` flag to allow you to see why the check failed. So, instead of a Boolean as a result there'd be some `ModulusCheckResult` object that holds the check status and optionally a string describing the result.

I'm just starting my working day so I'll let this sit in my subconscious and see what that generates by this evening.

Really pleased you're using the library though! And grateful you've shared the difference between your use-case and my expected ones...

sourcejon commented:
Awesome, well I will leave this with you. 

This is the implementation we run with as it gives just enough flexibility to the calling system to decide if it should strictly follow the rules or fail the modulus check when the sort code is not valid. This can be especially useful when pre-processing many thousands of accounts.

pauldambra commented:
My subconscious has just processes this... IIRC that a sortcode isn't in the list does not mean that it is an invalid sortcode. Only that it is not a sortcode which can be invalidated against an account number using the modulus checking algorithm

I think that if you want to check if a sortcode is 'real' then you need the EISCD

sourcejon commented:
Hey, been away for a few days.

Yeah your absolutely right, we just use this as a first pass before performing full checks which includes EISCD amongst other things. 

However, as I said before this is how we have interpreted the result for sortcode not in the list, so just though it would be nice to give other people the option of doing the same, as I am pretty confident most people who use your excellent c# program would like the flexibility to choose the outcome when the result is not in the list.

pauldambra commented:
Thanks for the prompt! It's great to see that people are using the library :) 

I'll add the start of a branch for explain mode in short order...
